assembly/src/assembler/module_provider.rs

```diff
@@ -22,6 +23,11 @@ impl ModuleProvider {
         self.procedures.get(id).map(|i| &self.modules[*i])
     }
 
+    /// Returns an iterator over the modules.
+    pub fn module_iter(&self) -> Iter<Module> {
+        self.modules.iter()
+    }
```

Comment:
With the approach I proposed in the above comments, I think we may not need this change.